Add details about how the dynamic configuration update REST API works [DEX-115] (#1121)
hazelcast/hz-docs: docs/modules/maintain-cluster/pages/enterprise-rest-api.adoc

  "message": "CP Subsystem is not enabled!"
}
----

== Dynamic Configuration Update REST Endpoint

> You can use the `/hazelcast/rest/api/v1/config/update` REST endpoint to change the dynamic server configurations (see xref:configuration:dynamic-config.adoc[Dynamic Configuration for Members]). The dynamic configuration lets you change different configurations in addition to adding new configurations for Hazelcast data structures.

Reviewer: ```suggestion
You can use the `/hazelcast/rest/api/v1/config/update` REST endpoint to change dynamic server configurations (for more info, see xref:configuration:dynamic-config.adoc[Dynamic Configuration for Members]). With dynamic configuration you can change existing configurations dynamically, as well as add new configurations for Hazelcast data structures.
```

> The endpoint requires that you send a xml/yaml server configuration file with the needed changes. The response will be two lists in json format. The first list is the newly added configurations among the configs sent to the server, and another list of ignored configurations which were in the sent configuration but could not be applied at the server, possibly since they are not dynamic configuration options but static configurations.

Reviewer 2: I think we need to update the info about the ignored configurations.

After running the endpoint multiple times, I observed the following behaviors:
1. If I add the same dynamic config more than once but in different requests, duplicate config is placed in ignored list.
2. If I try to add duplicate dynamic configs within the single request, `400` response is returned with a message `"interpretation error: the resource is neither valid XML nor valid YAML"`
3. If I try to add static configs, they are not placed in ignored list but also they are not placed in added list as well. Response is like below:
```
{
  "addedConfigs": [],
  "ignoredConfigs": []
}
```

Author: Answers:

1. We have this already in the config. I made different sections to highlight this now.
2. I tried this duplicate map config and received 200 OK:
```
curl -X 'POST' \
  'http://localhost:49567/hazelcast/rest/api/v1/cluster/config/update' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJyb2xlcyI6IltcImFkbWluXCJdIiwiaWF0IjoxNzE3NTg5ODE3LCJleHAiOjE3MTc1OTA3MTd9.2VVyxXPiF5710qw0VDwUGEHYh52nSdrsuQfcwyOkBeM' \
  -H 'Content-Type: application/json' \
  -d '<hazelcast xmlns="http://www.hazelcast.com/schema/config">
                       <map name="duplicate-map">
                           <statistics-enabled>true</statistics-enabled>
                       </map>
                       <map name="duplicate-map">
                           <statistics-enabled>true</statistics-enabled>
                       </map>
        </hazelcast>'
```
And I got this response:
```
{
  "addedConfigs": [
    {
      "sectionName": "map",
      "configName": "duplicate-map"
    }
  ],
  "ignoredConfigs": []
}
```

Can you share the example in your case? May be your XML is not correct.

3. Added this example.

Please see 7b5f64f375d28d2a34bd0ea0a0026ca39f43a406 for changes as suggested.

Author: You are correct. I trued this yaml file to start the server 
```yaml
hazelcast:
  map:
    my-map-yaml:
      in-memory-format: BINARY
      statistics-enabled: true
      backup-count: 2
    my-map-yaml:
      in-memory-format: BINARY
      statistics-enabled: true
      backup-count: 2
```
And the server fails with this message:
```
INFO: Loading configuration '/Users/ihsan/Desktop/hazelcast.yml' from System property 'hazelcast.config'
Jun 06, 2024 11:41:58 AM com.hazelcast.internal.config.AbstractConfigLocator
INFO: Using configuration file at /Users/ihsan/Desktop/hazelcast.yml
Exception in thread "main" com.hazelcast.config.InvalidConfigurationException: Invalid YAML configuration
	at com.hazelcast.config.YamlConfigBuilder.parseAndBuildConfig(YamlConfigBuilder.java:152)
	at com.hazelcast.config.YamlConfigBuilder.build(YamlConfigBuilder.java:138)
	at com.hazelcast.config.YamlConfigBuilder.build(YamlConfigBuilder.java:131)
	at com.hazelcast.config.Config.loadFromFile(Config.java:297)
	at com.hazelcast.config.Config.load(Config.java:261)
	at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance(HazelcastInstanceFactory.java:132)
	at com.hazelcast.core.Hazelcast.newHazelcastInstance(Hazelcast.java:95)
	at com.hazelcast.core.server.HazelcastMemberStarter.main(HazelcastMemberStarter.java:45)
Caused by: com.hazelcast.internal.yaml.YamlException: An error occurred while loading and parsing the YAML stream
	at com.hazelcast.internal.yaml.YamlLoader.load(YamlLoader.java:71)
	at com.hazelcast.config.YamlConfigBuilder.parseAndBuildConfig(YamlConfigBuilder.java:150)
	... 7 more
Caused by: while constructing a mapping
 in reader, line 3, column 5:
        my-map-yaml:
        ^
found duplicate key my-map-yaml
 in reader, line 7, column 5:
        my-map-yaml:
        ^

	at org.snakeyaml.engine.v2.constructor.StandardConstructor.processDuplicateKeys(StandardConstructor.java:90)
	at org.snakeyaml.engine.v2.constructor.StandardConstructor.flattenMapping(StandardConstructor.java:70)
	at org.snakeyaml.engine.v2.constructor.StandardConstructor.constructMapping2ndStep(StandardConstructor.java:119)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructMapping(BaseConstructor.java:278)
	at org.snakeyaml.engine.v2.constructor.StandardConstructor$ConstructYamlMap.construct(StandardConstructor.java:203)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructObjectNoCheck(BaseConstructor.java:153)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructObject(BaseConstructor.java:133)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructMapping2ndStep(BaseConstructor.java:302)
	at org.snakeyaml.engine.v2.constructor.StandardConstructor.constructMapping2ndStep(StandardConstructor.java:120)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructMapping(BaseConstructor.java:278)
	at org.snakeyaml.engine.v2.constructor.StandardConstructor$ConstructYamlMap.construct(StandardConstructor.java:203)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructObjectNoCheck(BaseConstructor.java:153)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructObject(BaseConstructor.java:133)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructMapping2ndStep(BaseConstructor.java:302)
	at org.snakeyaml.engine.v2.constructor.StandardConstructor.constructMapping2ndStep(StandardConstructor.java:120)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructMapping(BaseConstructor.java:278)
	at org.snakeyaml.engine.v2.constructor.StandardConstructor$ConstructYamlMap.construct(StandardConstructor.java:203)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructObjectNoCheck(BaseConstructor.java:153)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructObject(BaseConstructor.java:133)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.construct(BaseConstructor.java:92)
	at org.snakeyaml.engine.v2.constructor.BaseConstructor.constructSingleDocument(BaseConstructor.java:79)
	at org.snakeyaml.engine.v2.api.Load.loadOne(Load.java:111)
	at org.snakeyaml.engine.v2.api.Load.loadFromInputStream(Load.java:123)
	at com.hazelcast.internal.yaml.YamlLoader.load(YamlLoader.java:68)
	... 8 more
```

We did not set the [allow duplicate keys option](https://javadoc.io/static/org.snakeyaml/snakeyaml-engine/2.7/org/snakeyaml/engine/v2/api/LoadSettingsBuilder.html#setAllowDuplicateKeys(boolean)) in the yaml engine `LoadSettingsBuilder` hence it is false and fails by default. If I set it to true the server starts similar to the XML duplicate config. Hence, there is a behaviour difference at the server side for duplicate keys.

> The endpoint requires that you send a xml/yaml server configuration file with the needed changes. The response will be two lists in json format. The first list is the newly added configurations among the configs sent to the server, and another list of ignored configurations which were in the sent configuration but could not be applied at the server, possibly since they are not dynamic configuration options but static configurations.

Reviewer: ```suggestion
The endpoint requires that you send a XML/YAML server configuration file with the changes required. The response will be two lists in JSON format. The first list includes the newly added configuration among the configurations sent to the server. The second list includes ignored configurations which were in the sent configuration list but could not be applied at the server (examples of these could include static rather than dynamic configurations, which can not therefore be applied dynamically, or duplicate configurations already added).
```

> Let's now see how you can add a new `MapConfig` using this REST endpoint and analyse its response.

Reviewer: ```suggestion
As an example, let's add a new `MapConfig` using this REST endpoint, and analyze its response.
```

=== Add New MapConfig
Here is an example POST request for adding a new `MapConfig` for my-map:

[source,shell]
----
curl -X 'POST' \
  'http://localhost:8443/hazelcast/rest/api/v1/config/update' \
  -H 'Content-Type: text/plain' \
  -d '

>         <hazelcast xmlns="http://www.hazelcast.com/schema/config">

Reviewer 2: Can you also add yaml example?

                       <map name="my-map">
                           <in-memory-format>BINARY</in-memory-format>
                           <statistics-enabled>true</statistics-enabled>
                           <backup-count>2</backup-count>
                       </map>
        </hazelcast>
        '
----

> We send a request to create a new config for the new map to be created named `my-map`. The in-memory format is set to BINARY while the statistics is enabled and the backup count is set to 2.

Reviewer: ```suggestion
This request creates a configuration for a new map named `my-map`. The in-memory format is set to `BINARY`, while statistics are enabled and the backup count is set to `2`.
```

> When we send this request we receive the response body similar to this:

Reviewer: ```suggestion
In response to this request we should receive a response body such as:
```

[source,json]
----
{
  "addedConfigs": [
    {
      "sectionName": "map",
      "configName": "my-map"
    }
  ],
  "ignoredConfigs": []
}
----

> As you can observe from the response, a new map config is added for map with name my-map and all the requested dynamic configurations are applied.

Reviewer: ```suggestion
As the response indicates, a new map config is added to the `map` section with the name `my-map` and the requested dynamic configurations are applied.
```

> Now, if you like to add the config with the same map name and a different configuration, then you get response code 400 meaning an invalid configuration is provided since you can not change an existing map configuration. Let's try this:

Reviewer: ```suggestion
If you try to add a configuration with an already existing map name, you will receive a 400 response code which indicates an invalid configuration, as you cannot change an existing map configuration. Let's look at the code and response for this kind of example:
```

[source,shell]
----
curl -X 'POST' \
  'http://localhost:8443/hazelcast/rest/api/v1/config/update' \
  -H 'Content-Type: text/plain' \
  -d '
        <hazelcast xmlns="http://www.hazelcast.com/schema/config">
                       <map name="my-map">
                           <statistics-enabled>false</statistics-enabled>
                       </map>
        </hazelcast>
        '
----

> The response has the response code 400 and the error message why it was invalid configuration update request as follows:

Reviewer: ```suggestion
The response has a `statusCode` of 400, and the error message adds further detail:
```

}
----

> If you look carefully in the message, it says that "there is already a conflicting configuration" with the same map name.

Reviewer: ```suggestion
Within the response body you can see the explanation "as there is already a conflicting configuration 'MapConfig{name='my-map'".
```

> As you can see, you can not try adding a map config with the same map name but you can send an update request with the same configuration and map name, and you will observe that this update will be ignored since there is no change. Let's try the request again:

Reviewer: ```suggestion
As detailed above, you cannot add a map configuration with the same map name. If you send an update request with the same configuration and map name, this request will be ignored as there is no change. Let's look at this as an example:
```

Author: applied at aca85ecd19593850a288c42395c4db337fda2cab

[source,shell]
----
curl -X 'POST' \
  'http://localhost:8443/hazelcast/rest/api/v1/config/update' \
  -H 'Content-Type: text/plain' \
  -d '
        <hazelcast xmlns="http://www.hazelcast.com/schema/config">
                       <map name="my-map">
                           <in-memory-format>BINARY</in-memory-format>
                           <statistics-enabled>true</statistics-enabled>
                           <backup-count>2</backup-count>
                       </map>
        </hazelcast>
        '
----

> We receive a response code OK (200) and the following body:

Reviewer: ```suggestion
This request receives an `OK (200)` response code and the following body:
```

[source,json]
----
{
  "addedConfigs": [],
  "ignoredConfigs": [
    {
      "sectionName": "map",
      "configName": "my-map"
    }
  ]
}
----

> As you see from the response, no new configs added and the provided config is ignored.

Reviewer: ```suggestion
As the response confirms, no configs are added and the provided config is ignored.
```

> Whether a dynamic configuration can be applied or not depends on the request. For some updates, it is allowed to change configuration parameters of an existing configuration while for some as above you are not allowed to change it.

Reviewer: Can we provide more detail on when/if configuration changes are allowed? Is there certain parameters that can be changed or some logic to it? Otherwise it's just trial and error for user.

Author: @oliverhowell I refer to the main section in ref manual about the dynamic configuration feature. I also added the sentence like this: 
`Please see the xref:configuration:dynamic-config.adoc[Dynamic Configuration for Members] section on details about which configuration parameters can be changed dynamically and which can not be changed.
`

We are not really trying to explain the dynamic configuration feature itself here but how you can use it using the REST API. Hence, if there are missing information on dynamic configuration options, then it should be updated on the referred feature page. It is not in this document scope but surely we can improve that side as well. I tried to capture some common cases and what the user may see as the response but I do not aim to catch all the details of dynamic config in this REST API section.

Reviewer: Ah ok, makes sense. Maybe as a suggestion then we could add the 'for more info, see Dynamic Config' note to the end of this paragraph, you know the one we already use higher up the page?

Author: sure. changed as suggested at 624c767dd8680a5d8588518814ded2609bab7b11

}
----

== Dynamic Configuration Update REST Endpoint
You can use the `/hazelcast/rest/api/v1/config/update` REST endpoint to change dynamic server configurations (for more info, see xref:configuration:dynamic-config.adoc[Dynamic Configuration for Members]). With dynamic configuration you can change existing configurations dynamically, as well as add new configurations for Hazelcast data structures.

The endpoint requires that you send a XML/YAML server configuration file with the changes required. The response will be two lists in JSON format. The first list includes the newly added configuration among the configurations sent to the server. The second list includes ignored configurations which were in the sent configuration list but could not be applied at the server (examples of these could include static rather than dynamic configurations, which can not therefore be applied dynamically, or duplicate configurations already added).

Whether a dynamic configuration can be applied or not depends on the request. For some updates, it is allowed to change configuration parameters of an existing configuration while for some others you are not allowed to change the configuration.

> Let's now see different update requests and their results.

Reviewer: ```suggestion
Let's walk through some example requests and responses.
```
